### DISCUSSION TITLE
pymongo cursor timeouts - maybe

### DISCUSSION THREAD

**Discussion by pavlis** [mspass-team/mspass#124] [Design & Development]
I have encountered the problem I think is the same as that on noted in [this issue page](https://github.com/man-group/arctic/issues/439) with a package using pymongo.   

I am not certain this is the problem, but the only test I can think of is to try the same test multiple times and see it the failure point varies.  I can, of course, also use the solution that page points out and see if the problem goes away.

**Comment by wangyinz** [comment]
By solution do you mean the `cursorTimeoutMillis` option? I guess if your iteration over the cursor contains a long processing, it can timeout. This is something we should note in our documentation for sure.

**Comment by pavlis** [comment]
The above solution was not generic.  It was a special feature of the package that issue page was actually directed at.  i.e. I think it is a package like MsPASS that uses MongoDB. 

The solution I did find that seems to have solved the problem is changing the find call to this:
```
cursor=dbwf.find({},no_cursor_timeout = True)
```
The warning we must give our users that time you create a loop wherein the loop will be active for more than 10 minutes (my reading says that is the default cursor timeout period) they should add the no_cursor_timeout=True arg to a find operator that generates a cursor object.   My MongoDB book has a funny name for cursors created that way:  immortal cursors.

**Comment by wangyinz** [comment]
Cool! That is definitely a much better solution than to change the global config of the mongo server (i.e., the `cursorTimeoutMillis` option). Actually, I guess it only matters to a sequential job. Our `read_distributed_data` function will not cause a timeout as it reads in the `_id` immediately.

**Comment by pavlis** [comment]
No, I don't think that is necessarily true about parallel jobs.  It completely depends upon how spark and mongo interact on this issue.  From what I read the issue is pretty simple.  We are on the client side with pymongo.  When you create a cursor object it means a thread is launched (I assume) on the serve side to handle the conversation with the client.  That thread has an inactivity timeout that defaults to 10 minutes.   Any connnection where the cursor isn't forced to be update frequency is subject to this problem.  For a normal transaction even reading through a large collection that would rarely be an issue as the expection, I think, is that the cursor is run in a loop that hits a lot of documents fast.  The problem here is the cache is large enough and the loop contents are slow enough that the cursor times out before the cache is empty.  When it needs to refresh because the cache is almost empty you then get the cursor error because the server dropped the connection.  Not at all sure a parallel job will not be subject to the same problem, but we should worry about cases where it might break.  I don't understand the spark interaction well enough to know if this could actually happen. -

**Reply by haruming** [reply]
Hi Gary, I don't understand why "the cache is large enough and the loop contents are slow enough that the cursor times out before the cache is empty" will cause this problem. Do you mean there is a gap that the cursor is inactive for a long time then triggers the error?

**Comment by pavlis** [comment]
That happened in my test because the job had this basic structure:
```
curs=db.find()
for doc in curs:
   do some things
```
The problem is that when "do some things" takes a long the connection to the database server contained in the cursor object times out (I read the default timeout is 10 minutes).  The cache part is that a cursor loads documents into a fixed buffer size.   If the entire data set the job is trying to handle exceeds the buffer size the cursor will periodically need to ask the server to refresh its buffer.   It the time between requests is more than the timeout period you get this error.

That is my understanding of this problem from what I read in the docs.  Didn't preserve my search chain.

**Comment by pavlis** [comment]
I just hit another version of this problem that may require some digging to solve.   The problem has emerged from a final step in building a special arrival collection for the usarray data.  The collection is large by almost any measure at 2.7 million documents.   Nonetheless, similar, on the surface, algorithms seem to run much faster.  To give you an idea of how slow this is I did a rough timing test just done manually with a second terminal running find queries to the arrival collection to monitor how it was being update.   I found it was only sustaining around 80 updates per second.   That isn't awful, but for 2.7 million documents that explains why this has been running for hours.   The code I'm running is this (I'm pasting it  to include the code for the immortal cursor that has not yet been pushed to the repository):  
```
def set_source_id_from_evid(db,collection='arrival',use_immortal_cursor=False):
    """
    Sets source_id in arrivals by matching evid attributes in the
    source collection.   The concept here is that source_id is the
    unique key in MsPASS but evid is a foreign key defined when
    importing data from a CSS3.0 database.   The evid key can thus
    not be trusted so if we override it with a source_id we can
    guarantee it is unique.  This function will only work if the evid
    attribute is set in documents in the source collection.  Currently
    this can be done with a function called load_css30_sources.

    :param db:  MongoDB handle or MsPASS Database object managing these data.
    :param collection:  collection to search default is arrival, but any
      collection that has evid defined could be scanned.  The algorithm
      simply tries to match evid and updates documents it finds with
      the object id of the source record it matches.
    :param use_immortal_cursor:  If true the cursor in the update is made
      "immortal" meaning it won't time out.  This may be necessary if the
      arrival collection is large.
    :return: tuple of results with mixed content.  0 and 1 are integers
      defining number of documents processed and the number altered
      respectively.   2 is a dict keyed by evid with a value equal to the
      count of the number of documents found and altered with that evid.
      Component 3 is the dict of the complement to 2; evid keyed dict
      with the count of the number of documents encountered but which
      for which evid did not match any document in source.

    """
    dbarr=db[collection]
    dbsrc=db['source']
    if use_immortal_cursor:
        alldocs=dbarr.find({},no_cursor_timeout = True)
    else:
        alldocs=dbarr.find({})
    number_arrivals=0
    number_set=0
    evid_set=dict()
    not_set=dict()
    for doc in alldocs:
        if 'evid' in doc:
            evid=doc['evid']
            query={'evid' : evid}
            n=dbsrc.count_documents(query)
            if n==0:
                if evid in not_set:
                    nset=not_set[evid]
                    nset+=1
                    not_set[evid]=nset
                else:
                    not_set[evid]=1
            else:
                srcrec=dbsrc.find_one(query)
                source_id=srcrec['source_id']
                arroid=doc['_id']
                dbarr.update_one(
                        {'_id' : arroid},
                        {'$set' : {'source_id' : source_id}}
                )
                number_set += 1
                if evid in evid_set:
                    nset=evid_set[evid]
                    nset+=1
                    evid_set[evid]=nset
                else:
                    evid_set[evid]=1
        number_arrivals += 1
    return [number_arrivals,number_set,evid_set,not_set]
```
Much of the reason, I think, it is slow is a combination of two factors:
1.  Every single document is getting an update. 2.7 M updates is a lot not matter how you slice it.
2.  Each document in this algorithm results in 3 transaction:  count_documents on source, query to source, and then an update on arrival. I think a throttle may be embedded queries to source. Source is not huge, but the problem is it doesn't have an index for evid (the match key used in the query) and probably shouldn't because evid is foreign and maybe not even currently defined in our schema (that wouldn't really matter here though as this function predates the schema class development).   

There are two thing here that are worth discussing:
1.  I would argue we should consider adding the "use_immortal_cursor" argument for any database function that uses a cursor that might need that handling.  I don't know what functions that might effect right now, but it is worth noting as this will bite our users some day if we don't deal with it.  
2. This is a special case algorithm and fixing it for this one up problem is not necessary, but it would be good, I think, to know what is causing this.   Once it finishes (hopefully soon based on the high water mark at this moment) I'll export the database and upload it to stampede2 as planned.   Then I think I'll hack this code and add some timing functions to see where the bottleneck is.

**Reply by wangyinz** [reply]
It seems to me that the slow performance may be mostly due to the `query={'evid' : evid}`. Have you tried indexing `evid`? When querying a non-indexed field, I think mongodb will have to traverse the whole collection. Although the source collection may not be that big, the cost can add up.

The other thing I can think of is maybe the update performance. There is this subtlety in how mongodb ensures a write operation is performed (details [here](https://docs.mongodb.com/manual/core/write-performance/#journaling)). Maybe we should test turning off the journaling? We can either change the write concern setting of the arrival collection in this script (by setting w=0 with syntax similar to [this](https://stackoverflow.com/questions/41897158/pymongo-setting-w-for-collection-update-overrides-default-write-concern-options)), or turn off journaling globally in mongod ([example](https://docs.mongodb.com/manual/reference/program/mongod/#cmdoption-mongod-nojournal)).

**Comment by pavlis** [comment]
I made a mental math error in dealing with the counts I was getting back from my probing ipython terminal and it looks like I still have of the order of 4 hours before this crazy thing finishes.

I thought about adding evid as an index, but wasn't sure that was wise since not all documents have evid.  On the other hand, maybe that would be good as the index, I assume, would exclude any document where evid==None.   Don't think I can do that while it is running have it have an influence could I?   I don't want to crash this job after this long.   On the other hand, it demonstrates a need.  I should have the default behavior do an initial query that would replace 
```
alldocs=dbarr.find({},no_cursor_timeout = True)
```
with
```
alldocs=dbarr.find({'source_id' : {'$ne' : None}},no_cursor_timeout = True)
```
with no guarantee that is the right syntax.  Point is to limit the range to records where *source_id* is not yet set. 

With that model I think it would be good to add a *update_all=False* argument that would bypass the selection and force updating every document like this function is doing.  If you concur, I may kill this job, do that rewrite with adding a evid index and see if it speed up.

**Reply by wangyinz** [reply]
Well, I don't know. If it is gonna be 4 hours, maybe we can wait. But still, we will want to test out the performance difference with/without an index and turn on/off the journaling. It will be helpful to document them anyway.

**Comment by pavlis** [comment]
BTW:  correct query is this:  `dbadd.find({'source_id' : None})`.   I realized my probing ipython job was using just that construct.   The previous box, in fact, had the logic even reversed,

**Comment by pavlis** [comment]
I went ahead and did both of what was discussed above:  added the evid index and changed the function so it wouldn't start over again and only update unset values. 

Change in performance confirms your hypothesis pretty resoundingly:   went from 80 updates per second to around 1000 now.  Now this should be done in 20 or 30 minutes.  One more lesson in the use of indices in MongoDB.
